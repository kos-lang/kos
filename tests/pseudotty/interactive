# After displaying first command prompt, and after Kos queries terminal size
# via \e[6n, disable responses to \e[6n, so that further terminal queries
# go through COLUMNS env var.
#disable_cursor_pos

# Simple/basic test - print 42
print(42)\r

# ============================================================================
# Perform basic cursor movement on an empty command line

# Left
\e[D
\eOD
\x02

# Right
\e[C
\eOC
\x06

# Home
\e[H
\eOH
\e[1~
\e[7~
\x01

# End
\e[F
\eOF
\e[4~
\e[8~
\x05

# Backspace
\x7f
\x08

# Delete
\e[3~

# Ctrl-K
\x0B

# Ctrl-U
\x15

# Enter/Return
\r
\n

# ============================================================================
# Ctrl-C

abcd
\x03

# ============================================================================
# History

[123]
\e[A      # Up
# Modify print(42), but it should not affect history
\x7f\x7f  # Backspace x2
\e[B      # Down
\e[D      # Left
\x7f      # Backspace
\r        # RUN: [12]
xyz
\eOA      # Up
\x10      # Up
\r        # RUN: print(42)

# ============================================================================
# Insert with cursor movement

i
\e[D      # Left
r
\e[C      # Right
n
\eOD      # Left
\eOD      # Left
\eOD      # Left
\eOD      # Left
p
\eOC      # Right
\eOC      # Right
\eOC      # Right
\eOC      # Right
t(84)\r   # RUN: print(84)

\x10      # Up
\x7f      # Backspace

12345678)\r # RUN: print(8412345678)

# One more character than terminal width, will cause text scrolling

\x10      # Up
\x7F      # Backspace

9)\r      # RUN: print(84123456789)

# Insert characters in the middle and at the beginning

\x10      # Up
\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02 # Left x10
0
\x02\x02\x02\x02\x02\x02\x02\x02\x02\x02 # Left x10
\e[3~     # Delete
p\r

# ============================================================================
# Home/End cursor movement

es
\e[H      # Home
"t
\e[F      # End
t1"\r

es
\eOH      # Home
"t
\eOF      # End
t2"\r

es
\e[1~     # Home
"t
\e[4~     # End
t3"\r

es
\e[7~     # Home
"t
\e[8~     # End
t4"\r

es
\x01      # Home
"t
\x05      # End
t5"\r

BCDEFGHIJKLMNOPQRSTU
\e[H      # Home
\e[H      # Home
A
\e[F      # End
\e[F      # End
VWXY
\e[H      # Home
\e[H      # Home
"
\e[F      # End
\e[F      # End
Z"\r

# ============================================================================
# Exit

\x04
