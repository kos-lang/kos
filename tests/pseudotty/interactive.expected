R:Kos 0.1.0 interactive interpreter\r
R:\e[9999C\e[6n\r>
R: \e[0K
C:disable_cursor_pos
R:print(42)\r
R:42\r
R:\e[9999C\e[6n\r>
R: \e[0K\a\a\a\a\a\a\a\a\a\r
R:\r>
R: \e[0K\r
R:\r>
R: \e[0Kabcd\r
R:\r>
R: \e[0K\r
R:\r>
R: \e[0K[123]\r> print(42)\e[0K\r> print(42\e[0K\r> print(4\e[0K\r> [123]\e[0K\e[1D\r> [12]\e[0K\e[1D\r
R:[12]\r
R:\r>
R: \e[0Kxyz\r> [12]\e[0K\r> print(42)\e[0K\r
R:42\r
R:\r>
R: \e[0K\r> print(42)\e[0K\r> \e[0K\r> print(42)\e[0K\r> \e[0Ki\e[1D\r> ri\e[0K\e[1D\e[1Cn\e[1D\e[1D\e[1D\a\r> prin\e[0K\e[3D\e[1C\e[1C\e[1C\at(84)\r
R:84\r
R:\r>
R: \e[0K\r> print(84)\e[0K\r> print(84\e[0K12345678)\r
R:8412345678\r
R:\r>
R: \e[0K\r> print(8412345678)\e[0K\r> print(8412345678\e[0K9\r> rint(84123456789)\e[0K\r
R:84123456789\r
R:\r>
R: \e[0K\r> rint(84123456789)\e[0K\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\r> rint(840123456789)\e[10D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\r> print(840123456789\e[18D\a\r> rint(840123456789)\e[18D\r> print(840123456789\e[17D\r
R:840123456789\r
R:\r>
R: \e[0Kes\e[2D\r> "es\e[0K\e[2D\r> "tes\e[0K\e[2D\e[2Ct1"\r
R:"test1"\r
R:\r>
R: \e[0Kes\e[2D\r> "es\e[0K\e[2D\r> "tes\e[0K\e[2D\e[2Ct2"\r
R:"test2"\r
R:\r>
R: \e[0Kes\e[2D\r> "es\e[0K\e[2D\r> "tes\e[0K\e[2D\e[2Ct3"\r
R:"test3"\r
R:\r>
R: \e[0Kes\e[2D\r> "es\e[0K\e[2D\r> "tes\e[0K\e[2D\e[2Ct4"\r
R:"test4"\r
R:\r>
R: \e[0Kes\e[2D\r> "es\e[0K\e[2D\r> "tes\e[0K\e[2D\e[2Ct5"\r
R:"test5"\r
R:\r>
R: \e[0KBCDEFGHIJKLMNOPQR\r> CDEFGHIJKLMNOPQRS\e[0K\r> DEFGHIJKLMNOPQRST\e[0K\r> EFGHIJKLMNOPQRSTU\e[0K\r> BCDEFGHIJKLMNOPQRS\e[18D\r> ABCDEFGHIJKLMNOPQR\e[17D\r> EFGHIJKLMNOPQRSTU\e[0K\r> FGHIJKLMNOPQRSTUV\e[0K\r> GHIJKLMNOPQRSTUVW\e[0K\r> HIJKLMNOPQRSTUVWX\e[0K\r> IJKLMNOPQRSTUVWXY\e[0K\r> ABCDEFGHIJKLMNOPQR\e[18D\r> "ABCDEFGHIJKLMNOPQ\e[17D\r> IJKLMNOPQRSTUVWXY\e[0K\r> JKLMNOPQRSTUVWXYZ\e[0K\r> KLMNOPQRSTUVWXYZ"\e[0K\r
R:"ABCDEFGHIJKLMNOPQRSTUVWXYZ"\r
R:\r>
R: \e[0Kx"Hello"y\r> x"Hello"\e[0K\e[8D\e[1C\r> "Hello"\e[0K\e[7D\a\r
R:"Hello"\r
R:\r>
R: \e[0Kx"abc defghijklmn\r> "abc defghijklmno\e[0K\r> abc defghijklmnop\e[0K\r> bc defghijklmnopq\e[0K\r> c defghijklmnopqr\e[0K\r>  defghijklmnopqrs\e[0K\r> defghijklmnopqrst\e[0K\r> efghijklmnopqrst.\e[0K\r> fghijklmnopqrst.u\e[0K\r> ghijklmnopqrst.uv\e[0K\r> hijklmnopqrst.uvw\e[0K\r> ijklmnopqrst.uvwx\e[0K\r> jklmnopqrst.uvwxy\e[0K\r> klmnopqrst.uvwxyz\e[0K\r> lmnopqrst.uvwxyz"\e[0K\r> mnopqrst.uvwxyz"y\e[0K\r> x"abc defghijklmno\e[18D\e[1C\r> "abc defghijklmnop\e[18D\r> mnopqrst.uvwxyz"y\e[0K\r> mnopqrst.uvwxyz"\e[0K\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\r> lmnopqrst.uvwxyz"\e[0K\e[17D\r> klmnopqrst.uvwxyz"\e[18D\r> jklmnopqrst.uvwxyz\e[18D\r> ijklmnopqrst.uvwxy\e[18D\r> hijklmnopqrst.uvwx\e[18D\r> ghijklmnopqrst.uvw\e[18D\r> fghijklmnopqrst.uv\e[18D\r> efghijklmnopqrst.u\e[18D\r> defghijklmnopqrst.\e[18D\r> defghijklmnopqrst.\e[18D\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\r> efghijklmnopqrst.u\e[1D\r> efghijklmnopqrstuv\e[2D\r
R:"abcdefghijklmnopqrstuvwxyz"\r
R:\r>
R: \e[0Kx"Hello2"y\a\e[1D\r> x"Hello2"\e[0K\a\e[9D\r> "Hello2"\e[0K\e[8D\r
R:"Hello2"\r
R:\r>
R: \e[0Kx"abc defghijklmn\r> "abc defghijklmno\e[0K\r> abc defghijklmnop\e[0K\r> bc defghijklmnopq\e[0K\r> c defghijklmnopqr\e[0K\r>  defghijklmnopqrs\e[0K\r> defghijklmnopqrst\e[0K\r> efghijklmnopqrst.\e[0K\r> fghijklmnopqrst.u\e[0K\r> ghijklmnopqrst.uv\e[0K\r> hijklmnopqrst.uvw\e[0K\r> ijklmnopqrst.uvwx\e[0K\r> jklmnopqrst.uvwxy\e[0K\r> klmnopqrst.uvwxyz\e[0K\r> lmnopqrst.uvwxyz"\e[0K\r> mnopqrst.uvwxyz"y\e[0K\r> x"abc defghijklmno\e[18D\r> "abc defghijklmnop\e[18D\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\e[1C\r> abc defghijklmnopq\e[1D\r> bc defghijklmnopqr\e[1D\r> c defghijklmnopqrs\e[1D\r>  defghijklmnopqrst\e[1D\r> defghijklmnopqrst.\e[1D\r> defghijklmnopqrstu\e[1D\r> lmnopqrstuvwxyz"y\e[0K\e[1D\r> lmnopqrstuvwxyz"\e[0K\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\e[1D\r> klmnopqrstuvwxyz"\e[0K\e[17D\r> jklmnopqrstuvwxyz"\e[18D\r> ijklmnopqrstuvwxyz\e[18D\r> hijklmnopqrstuvwxy\e[18D\r> ghijklmnopqrstuvwx\e[18D\r> fghijklmnopqrstuvw\e[18D\r> efghijklmnopqrstuv\e[18D\r> defghijklmnopqrstu\e[18D\r>  defghijklmnopqrst\e[18D\r> defghijklmnopqrstu\e[18D\r
R:"abcdefghijklmnopqrstuvwxyz"\r
R:\r>
R: \e[0K123xyz\e[1D\e[1D\e[1D\r> 123yz\e[0K\e[2D\r> 123z\e[0K\e[1D\r> 123\e[0K\a\r
R:123\r
R:\r>
R: \e[0K123xyz\e[1D\e[1D\e[1D\e[0K\r
R:123\r
R:\r>
R: \e[0K\r> 123\e[0K\r> 123\e[0K\r> klmnopqrstuvwxyz"\e[0K\r> "abcdefghijklmnopq\e[18D\e[1C\e[1C\e[1C\e[0K"\r
R:"ab"\r
R:\r>
R: \e[0K\r> "ab"\e[0K\r> 123\e[0K\r> 123\e[0K\r> klmnopqrstuvwxyz"\e[0K\e[1D\e[1D\e[1D\e[1D\e[0K"\r
R:"abcdefghijklmnopqrstuvw"\r
R:\r>
R: \e[0K1234567890\r> \e[0K"abcd"\r
R:"abcd"\r
R:\r>
R: \e[0K12345678901234567\r> 23456789012345678\e[0K\r> 34567890123456789\e[0K\r> 45678901234567890\e[0K\r> 56789012345678901\e[0K\r> 67890123456789012\e[0K\r> 78901234567890123\e[0K\r> 89012345678901234\e[0K\r> 90123456789012345\e[0K\r> 01234567890123456\e[0K\r> 12345678901234567\e[0K\r> 23456789012345678\e[0K\r> 34567890123456789\e[0K\r> 45678901234567890\e[0K\r> 56789012345678901\e[0K\r> 67890123456789012\e[0K\r> 78901234567890123\e[0K\r> 89012345678901234\e[0K\r> 90123456789012345\e[0K\r> 01234567890123456\e[0K\r> 12345678901234567\e[0K\r> 23456789012345678\e[0K\r> 34567890123456789\e[0K\r> 45678901234567890\e[0K\r> \e[0K"abcd"\r
R:"abcd"\r
R:\r>
R: \e[0K\a      Abc      45\r>      Abc      456\e[0K\r>     Abc      456 \e[0K\r>    Abc      456  \e[0K\r>   Abc      456   \e[0K\r>  Abc      456    \e[0K\r> Abc      456     \e[0K\r> bc      456      \e[0K\r> c      456      7\e[0K\r>       456      78\e[0K\r>      456      789\e[0K\r>     456      789 \e[0K\r>    456      789  \e[0K\r>   456      789   \e[0K\r>  456      789    \e[0K\r> 456      789     \e[0K\r> 56      789      \e[0K\r> 6      789      0\e[0K\r> 6      789      \e[0K\e[9D\r> 456      789      \e[18D\r> Abc      456      \e[18D\e[1C\e[1C\r> c      456      78\e[18D\e[1C\e[1C\e[1C\e[1C\r>    456      789   \e[18D\e[6C\r>    456+      789  \e[11D\r
R:1245\r
R:\r>
R: \e[0K      123      45\r>      123      456\e[0K\r>     123      456 \e[0K\r>    123      456  \e[0K\r>   123      456   \e[0K\r>  123      456    \e[0K\r> 123      456     \e[0K\r> 23      456      \e[0K\r> 3      456      7\e[0K\r>       456      78\e[0K\r>      456      789\e[0K\r>     456      789 \e[0K\r>    456      789  \e[0K\r>   456      789   \e[0K\r>  456      789    \e[0K\r> 456      789     \e[0K\r> 56      789      \e[0K\r> 6      789      0\e[0K\r> 6      789      \e[0K\r>       123      456\e[18D\r>       456      789\e[18D\r> 0      456      78\e[17D\r> 0+      456      7\e[16D\e[9C\r> 0+      456-      \e[6D\r
R:-333\r
R:\r>
R: \e[0K123     456\e[11D\e[3C\r>      456\e[0K\e[8D\r
R:456\r
R:\r>
R: \e[0Kp\ar\ai\an\at\a(84)\r
R:84\r
R:\r>
R: \e[0K256->\e[H\e[2J\r> 256->\e[0Kprint\r
R:256\r
R:\r>
R: \e[0K\a"\r> "\xc2\x8f\e[0K\r> "\xc2\x8f\xe8\x82\x83\e[0K"\e[1D\e[1D\r> "\xe8\x82\x83"\e[0K\e[2D\r
R:"\x{8083}"\r
R:\r>
R: \e[0K
EXIT:0
